File: backend/app/schemas/company.py
from pydantic import BaseModel, ConfigDict
from typing import Optional, Dict, Any, List
from datetime import datetime


class CompanyBase(BaseModel):
    """Base company fields matching the Supabase companies table."""
    model_config = ConfigDict(from_attributes=True, extra="allow")

    name: str
    sector: Optional[str] = None
    stage: Optional[str] = None
    status: Optional[str] = None
    description: Optional[str] = None
    website: Optional[str] = None
    headquarters: Optional[str] = None
    business_model: Optional[str] = None
    revenue_model: Optional[str] = None
    kpi_framework: Optional[str] = None
    funding_stage: Optional[str] = None
    founded_year: Optional[int] = None
    ai_first: Optional[bool] = None
    ai_category: Optional[str] = None
    category: Optional[str] = None


class CompanyMinimal(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="allow")

    id: str
    name: str
    current_arr_usd: Optional[float] = None
    sector: Optional[str] = None
    current_valuation_usd: Optional[float] = None
    founded_year: Optional[int] = None


class Company(CompanyBase):
    """Full company schema matching Supabase companies table columns."""
    id: str
    organization_id: Optional[str] = None
    fund_id: Optional[str] = None

    # Revenue metrics
    current_arr_usd: Optional[float] = None
    current_mrr_usd: Optional[float] = None
    revenue_growth_monthly_pct: Optional[float] = None
    revenue_growth_annual_pct: Optional[float] = None

    # Valuation
    current_valuation_usd: Optional[float] = None
    last_valuation_usd: Optional[float] = None

    # Burn & runway
    burn_rate_monthly_usd: Optional[float] = None
    runway_months: Optional[float] = None
    cash_in_bank_usd: Optional[float] = None

    # Funding
    total_raised: Optional[float] = None
    total_invested_usd: Optional[float] = None
    amount_raised: Optional[Any] = None  # JSONB in DB
    round_size: Optional[Any] = None  # JSONB in DB
    quarter_raised: Optional[str] = None
    last_funding_date: Optional[str] = None

    # Growth & efficiency
    growth_rate: Optional[float] = None
    gross_margin: Optional[float] = None
    employee_count: Optional[int] = None
    tam: Optional[float] = None
    current_option_pool_bps: Optional[float] = None

    # Funnel & sourcing
    funnel_status: Optional[str] = None
    thesis_match_score: Optional[float] = None
    recommendation_reason: Optional[Any] = None  # JSONB
    added_to_watchlist_at: Optional[str] = None
    watchlist_priority: Optional[str] = None

    # Term sheet
    term_sheet_sent_at: Optional[str] = None
    term_sheet_status: Optional[str] = None
    term_sheet_expiry_date: Optional[str] = None

    # Investment tracking
    first_investment_date: Optional[str] = None
    latest_investment_date: Optional[str] = None
    ownership_percentage: Optional[float] = None
    investment_lead: Optional[str] = None

    # Exit
    exit_date: Optional[str] = None
    exit_type: Optional[str] = None
    exit_value_usd: Optional[float] = None
    exit_multiple: Optional[float] = None

    # Customer segments
    customer_segment_enterprise_pct: Optional[float] = None
    customer_segment_midmarket_pct: Optional[float] = None
    customer_segment_sme_pct: Optional[float] = None

    # PWERM
    has_pwerm_model: Optional[bool] = None
    latest_pwerm_run_at: Optional[str] = None
    pwerm_scenarios_count: Optional[int] = None

    # Updates tracking
    latest_update: Optional[str] = None
    latest_update_date: Optional[str] = None
    update_frequency_days: Optional[int] = None
    last_contacted_date: Optional[str] = None

    # Field update timestamps
    cash_updated_at: Optional[str] = None
    burn_rate_updated_at: Optional[str] = None
    runway_updated_at: Optional[str] = None
    revenue_updated_at: Optional[str] = None
    gross_margin_updated_at: Optional[str] = None

    # Location & flexible data
    location: Optional[Any] = None  # JSONB
    extra_data: Optional[Dict[str, Any]] = None
    currency_mix: Optional[Dict[str, float]] = None
    metadata: Optional[Dict[str, Any]] = None

    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class CompanyCreate(CompanyBase):
    pass


class CompanyUpdate(BaseModel):
    model_config = ConfigDict(extra="allow")

    name: Optional[str] = None
    sector: Optional[str] = None
    stage: Optional[str] = None
    status: Optional[str] = None
    current_arr_usd: Optional[float] = None
    current_valuation_usd: Optional[float] = None
    burn_rate_monthly_usd: Optional[float] = None
    growth_rate: Optional[float] = None
    total_raised: Optional[float] = None
    extra_data: Optional[Dict[str, Any]] = None
    metadata: Optional[Dict[str, Any]] = None

File: frontend/src/types/company.ts
export interface Company {
    id: string;
    organization_id?: string;
    fund_id?: string;
    name: string;
    website?: string;
    sector?: string;
    revenue_model?: string;
    kpi_framework?: string;
    current_arr_usd?: number;
    current_mrr_usd?: number;
    revenue_growth_monthly_pct?: number;
    revenue_growth_annual_pct?: number;
    burn_rate_monthly_usd?: number;
    runway_months?: number;
    current_option_pool_bps?: number;
    location?: Record<string, any>;
    status: string;
    created_at: string;
    updated_at: string;
    round_size?: Record<string, any>;
    amount_raised?: Record<string, any>;
    quarter_raised?: any;
    funnel_status: string;
    thesis_match_score?: number;
    recommendation_reason?: Record<string, any>;
    added_to_watchlist_at?: string;
    watchlist_priority?: string;
    term_sheet_sent_at?: string;
    term_sheet_status?: string;
    term_sheet_expiry_date?: string;
    first_investment_date?: string;
    latest_investment_date?: string;
    total_invested_usd?: number;
    ownership_percentage?: number;
    exit_date?: string;
    exit_type?: string;
    exit_value_usd?: number;
    exit_multiple?: number;
    customer_segment_enterprise_pct?: number;
    customer_segment_midmarket_pct?: number;
    customer_segment_sme_pct?: number;
    latest_update?: string;
    latest_update_date?: string;
    update_frequency_days?: number;
    has_pwerm_model?: boolean;
    latest_pwerm_run_at?: string;
    pwerm_scenarios_count?: number;
    gross_margin?: number;
    category?: string;
    ai_first?: boolean;
    business_model?: string;
    // Portfolio report fields
    cash_in_bank_usd?: number;
    investment_lead?: string;
    last_contacted_date?: string;
    // Field update timestamps
    cash_updated_at?: string;
    burn_rate_updated_at?: string;
    runway_updated_at?: string;
    revenue_updated_at?: string;
    gross_margin_updated_at?: string;
    // Recent migration fields
    current_valuation_usd?: number;
    extra_data?: Record<string, any>;
    currency_mix?: Record<string, number>;
  }

export interface PortfolioCompany {
  id: string;
  company_id: string;
  company_name: string;
  investment_date: string;
  investment_amount_usd: number;
  ownership_percentage: number;
  current_valuation_usd: number;
  status: 'active' | 'exited' | 'written_off';
  exit_date?: string;
  exit_value_usd?: number;
  exit_multiple?: number;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface PWERMScenario {
  scenario_name: string;
  probability: number;
  exit_value_usd: number;
  exit_multiple: number;
  exit_year: number;
  exit_type: string;
  assumptions: Record<string, any>;
  description: string;
  waterfall_adjusted_value: number;
  dilution_adjusted_ownership: number;
}

export interface WaterfallAssumptions {
  preferred_return_rate: number;
  catch_up_percentage: number;
  carried_interest_rate: number;
  management_fee_rate: number;
  hurdle_rate: number;
  gp_commitment_percentage: number;
}

export interface OptionsDilutionAssumptions {
  current_esop_percentage: number;
  future_esop_grants_per_year: number;
  vesting_schedule_years: number;
  cliff_percentage: number;
  anti_dilution_protection: boolean;
  future_funding_rounds: number;
  dilution_per_round: number;
}

export interface PWERMResult {
  company_id: string;
  company_name: string;
  analysis_date: string;
  total_invested_usd: number;
  ownership_percentage: number;
  scenarios: PWERMScenario[];
  expected_return_usd: number;
  expected_multiple: number;
  expected_irr: number;
  risk_adjusted_return: number;
  confidence_interval_95: [number, number];
  scenario_count: number;
  analysis_metadata: Record<string, any>;
  waterfall_assumptions: WaterfallAssumptions;
  dilution_assumptions: OptionsDilutionAssumptions;
  waterfall_adjusted_returns?: Record<string, number>;
  dilution_adjusted_ownership: number;
}

export interface PWERMPlaygroundConfig {
  company_id: string;
  base_assumptions: {
    graduation_rate: number;
    arr_growth_rate: number;
    exit_multiple_range: [number, number];
    dilution_per_round: number;
    time_to_exit_years: number;
    sector_multiplier: number;
    market_conditions: 'bull' | 'neutral' | 'bear';
  };
  waterfall_assumptions: WaterfallAssumptions;
  dilution_assumptions: OptionsDilutionAssumptions;
  ai_monitoring?: boolean;
}